File: frontend/src/app/features/forms/form.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { FormRequest, FormResponse } from '../../shared/models/form.model';

/**
 * HTTP access to the /api/forms endpoints.
 */
@Injectable({ providedIn: 'root' })
export class FormService {
  private readonly http = inject(HttpClient);

  private readonly baseUrl = `${environment.apiUrl}/forms`;

  /**
   * Reads every form, active or not.
   *
   * @returns list of forms
   */
  findAll(): Observable<FormResponse[]> {
    return this.http.get<FormResponse[]>(this.baseUrl);
  }

  /**
   * Reads the forms that can currently be answered. This endpoint is public.
   *
   * @returns list of active forms
   */
  findActive(): Observable<FormResponse[]> {
    return this.http.get<FormResponse[]>(`${this.baseUrl}/active`);
  }

  /**
   * Reads a single form with its questions.
   *
   * @param id identifier of the form
   * @returns the form
   */
  findById(id: number): Observable<FormResponse> {
    return this.http.get<FormResponse>(`${this.baseUrl}/${id}`);
  }

  /**
   * Creates a form together with its questions.
   *
   * @param request data typed in the editor
   * @returns the created form
   */
  create(request: FormRequest): Observable<FormResponse> {
    return this.http.post<FormResponse>(this.baseUrl, request);
  }

  /**
   * Updates a form and its questions.
   *
   * @param id      identifier of the form
   * @param request data typed in the editor
   * @returns the updated form
   */
  update(id: number, request: FormRequest): Observable<FormResponse> {
    return this.http.put<FormResponse>(`${this.baseUrl}/${id}`, request);
  }

  /**
   * Deactivates a form so it is no longer offered to the employees.
   *
   * @param id identifier of the form
   * @returns an empty response
   */
  deactivate(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  /**
   * Sends a form again to the employees of a company, as the yearly follow up.
   *
   * @param id        identifier of the form
   * @param companyId identifier of the company to notify
   * @returns an empty response
   */
  resendAnnually(id: number, companyId: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/${id}/resend/${companyId}`, {});
  }
}

File: backend/src/main/java/com/mgs/ergomanager/service/FormService.java
package com.mgs.ergomanager.service;

import com.mgs.ergomanager.dto.form.FormRequest;
import com.mgs.ergomanager.dto.form.FormResponse;
import java.util.List;

/**
 * Management of the self evaluation forms and their annual resend.
 */
public interface FormService {

    /**
     * Returns every form, active or not.
     *
     * @return list of forms
     */
    List<FormResponse> findAll();

    /**
     * Returns the forms that can currently be answered.
     *
     * @return list of active forms
     */
    List<FormResponse> findActive();

    /**
     * Returns a single form with its questions.
     *
     * @param id identifier of the form
     * @return the form
     */
    FormResponse findById(Long id);

    /**
     * Creates a form together with its questions.
     *
     * @param request data of the form
     * @return the created form
     */
    FormResponse create(FormRequest request);

    /**
     * Updates a form and its questions.
     *
     * @param id      identifier of the form
     * @param request new data of the form
     * @return the updated form
     */
    FormResponse update(Long id, FormRequest request);

    /**
     * Deactivates a form so it is no longer offered to the employees.
     *
     * @param id identifier of the form
     */
    void deactivate(Long id);

    /**
     * Sends a form again to the employees of a company, which is the yearly
     * follow up required by MGS.
     *
     * @param formId    identifier of the form to resend
     * @param companyId identifier of the company to notify
     */
    void resendAnnually(Long formId, Long companyId);
}
